File: java2csharp/java/Sinch/Sinch.Video.Messaging/SinchService.java
package com.sinch.android.rtc.sample.messaging;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import android.util.Log;

import com.sinch.android.rtc.*;
import com.sinch.android.rtc.messaging.MessageClientListener;
import com.sinch.android.rtc.messaging.WritableMessage;

public class SinchService extends Service {

    private static final String APP_KEY = "enter-application-key";
    private static final String APP_SECRET = "enter-application-secret";
    private static final String ENVIRONMENT = "sandbox.sinch.com";

    private static final String TAG = SinchService.class.getSimpleName();

    private final SinchServiceInterface mServiceInterface = new SinchServiceInterface();

    private SinchClient mSinchClient = null;
    private StartFailedListener mListener;

    public class SinchServiceInterface extends Binder {

        public boolean isStarted() {
            return SinchService.this.isStarted();
        }

        public void startClient(String userName) {
            start(userName);
        }

        public void stopClient() {
            stop();
        }

        public void setStartListener(StartFailedListener listener) {
            mListener = listener;
        }

        public void sendMessage(String recipientUserId, String textBody) {
            SinchService.this.sendMessage(recipientUserId, textBody);
        }

        public void addMessageClientListener(MessageClientListener listener) {
            SinchService.this.addMessageClientListener(listener);
        }

        public void removeMessageClientListener(MessageClientListener listener) {
            SinchService.this.removeMessageClientListener(listener);
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public void onDestroy() {
        if (mSinchClient != null && mSinchClient.isStarted()) {
            mSinchClient.terminate();
        }
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return mServiceInterface;
    }

    private boolean isStarted() {
        return (mSinchClient != null && mSinchClient.isStarted());
    }

    public void sendMessage(String recipientUserId, String textBody) {
        if (isStarted()) {
            WritableMessage message = new WritableMessage(recipientUserId, textBody);
            mSinchClient.getMessageClient().send(message);
        }
    }

    public void addMessageClientListener(MessageClientListener listener) {
        if (mSinchClient != null) {
            mSinchClient.getMessageClient().addMessageClientListener(listener);
        }
    }

    public void removeMessageClientListener(MessageClientListener listener) {
        if (mSinchClient != null) {
            mSinchClient.getMessageClient().removeMessageClientListener(listener);
        }
    }

    private void start(String userName) {
        if (mSinchClient == null) {
            mSinchClient = Sinch.getSinchClientBuilder().context(getApplicationContext()).userId(userName)
                    .applicationKey(APP_KEY)
                    .applicationSecret(APP_SECRET)
                    .environmentHost(ENVIRONMENT).build();

            mSinchClient.setSupportMessaging(true);
            mSinchClient.startListeningOnActiveConnection();

            mSinchClient.addSinchClientListener(new MySinchClientListener());
            mSinchClient.start();
        }
    }

    private void stop() {
        if (mSinchClient != null) {
            mSinchClient.terminate();
            mSinchClient = null;
        }
    }

    public interface StartFailedListener {

        void onStartFailed(SinchError error);

        void onStarted();
    }

    private class MySinchClientListener implements SinchClientListener {

        @Override
        public void onClientFailed(SinchClient client, SinchError error) {
            if (mListener != null) {
                mListener.onStartFailed(error);
            }
            mSinchClient.terminate();
            mSinchClient = null;
        }

        @Override
        public void onClientStarted(SinchClient client) {
            Log.d(TAG, "SinchClient started");
            if (mListener != null) {
                mListener.onStarted();
            }
        }

        @Override
        public void onClientStopped(SinchClient client) {
            Log.d(TAG, "SinchClient stopped");
        }

        @Override
        public void onLogMessage(int level, String area, String message) {
            switch (level) {
                case Log.DEBUG:
                    Log.d(area, message);
                    break;
                case Log.ERROR:
                    Log.e(area, message);
                    break;
                case Log.INFO:
                    Log.i(area, message);
                    break;
                case Log.VERBOSE:
                    Log.v(area, message);
                    break;
                case Log.WARN:
                    Log.w(area, message);
                    break;
            }
        }

        @Override
        public void onRegistrationCredentialsRequired(SinchClient client,
                ClientRegistration clientRegistration) {
        }
    }
}

File: java2csharp/csharp/Sinch/Sinch.Video.Messaging/SinchService.cs
namespace com.sinch.android.rtc.sample.messaging
{

	using Service = android.app.Service;
	using Intent = android.content.Intent;
	using Binder = android.os.Binder;
	using IBinder = android.os.IBinder;
	using Log = android.util.Log;

	using com.sinch.android.rtc;
	using MessageClientListener = com.sinch.android.rtc.messaging.MessageClientListener;
	using WritableMessage = com.sinch.android.rtc.messaging.WritableMessage;

	public class SinchService : Service
	{
		private bool InstanceFieldsInitialized = false;

		public SinchService()
		{
			if (!InstanceFieldsInitialized)
			{
				InitializeInstanceFields();
				InstanceFieldsInitialized = true;
			}
		}

		private void InitializeInstanceFields()
		{
			mServiceInterface = new SinchServiceInterface(this);
		}


		private const string APP_KEY = "enter-application-key";
		private const string APP_SECRET = "enter-application-secret";
		private const string ENVIRONMENT = "sandbox.sinch.com";

		private static readonly string TAG = typeof(SinchService).Name;

		private SinchServiceInterface mServiceInterface;

		private SinchClient mSinchClient = null;
		private StartFailedListener mListener;

		public class SinchServiceInterface : Binder
		{
			private readonly SinchService outerInstance;

			public SinchServiceInterface(SinchService outerInstance)
			{
				this.outerInstance = outerInstance;
			}


			public virtual bool Started
			{
				get
				{
					return outerInstance.Started;
				}
			}

			public virtual void startClient(string userName)
			{
				outerInstance.start(userName);
			}

			public virtual void stopClient()
			{
				outerInstance.stop();
			}

			public virtual StartFailedListener StartListener
			{
				set
				{
					outerInstance.mListener = value;
				}
			}

			public virtual void sendMessage(string recipientUserId, string textBody)
			{
				outerInstance.sendMessage(recipientUserId, textBody);
			}

			public virtual void addMessageClientListener(MessageClientListener listener)
			{
				outerInstance.addMessageClientListener(listener);
			}

			public virtual void removeMessageClientListener(MessageClientListener listener)
			{
				outerInstance.removeMessageClientListener(listener);
			}
		}

		public override void onCreate()
		{
			base.onCreate();
		}

		public override void onDestroy()
		{
			if (mSinchClient != null && mSinchClient.Started)
			{
				mSinchClient.terminate();
			}
			base.onDestroy();
		}

		public override IBinder onBind(Intent intent)
		{
			return mServiceInterface;
		}

		private bool Started
		{
			get
			{
				return (mSinchClient != null && mSinchClient.Started);
			}
		}

		public virtual void sendMessage(string recipientUserId, string textBody)
		{
			if (Started)
			{
				WritableMessage message = new WritableMessage(recipientUserId, textBody);
				mSinchClient.MessageClient.send(message);
			}
		}

		public virtual void addMessageClientListener(MessageClientListener listener)
		{
			if (mSinchClient != null)
			{
				mSinchClient.MessageClient.addMessageClientListener(listener);
			}
		}

		public virtual void removeMessageClientListener(MessageClientListener listener)
		{
			if (mSinchClient != null)
			{
				mSinchClient.MessageClient.removeMessageClientListener(listener);
			}
		}

		private void start(string userName)
		{
			if (mSinchClient == null)
			{
				mSinchClient = Sinch.SinchClientBuilder.context(ApplicationContext).userId(userName).applicationKey(APP_KEY).applicationSecret(APP_SECRET).environmentHost(ENVIRONMENT).build();

				mSinchClient.SupportMessaging = true;
				mSinchClient.startListeningOnActiveConnection();

				mSinchClient.addSinchClientListener(new MySinchClientListener(this));
				mSinchClient.start();
			}
		}

		private void stop()
		{
			if (mSinchClient != null)
			{
				mSinchClient.terminate();
				mSinchClient = null;
			}
		}

		public interface StartFailedListener
		{

			void onStartFailed(SinchError error);

			void onStarted();
		}

		private class MySinchClientListener : SinchClientListener
		{
			private readonly SinchService outerInstance;

			public MySinchClientListener(SinchService outerInstance)
			{
				this.outerInstance = outerInstance;
			}


			public override void onClientFailed(SinchClient client, SinchError error)
			{
				if (outerInstance.mListener != null)
				{
					outerInstance.mListener.onStartFailed(error);
				}
				outerInstance.mSinchClient.terminate();
				outerInstance.mSinchClient = null;
			}

			public override void onClientStarted(SinchClient client)
			{
				Log.d(TAG, "SinchClient started");
				if (outerInstance.mListener != null)
				{
					outerInstance.mListener.onStarted();
				}
			}

			public override void onClientStopped(SinchClient client)
			{
				Log.d(TAG, "SinchClient stopped");
			}

			public override void onLogMessage(int level, string area, string message)
			{
				switch (level)
				{
					case Log.DEBUG:
						Log.d(area, message);
						break;
					case Log.ERROR:
						Log.e(area, message);
						break;
					case Log.INFO:
						Log.i(area, message);
						break;
					case Log.VERBOSE:
						Log.v(area, message);
						break;
					case Log.WARN:
						Log.w(area, message);
						break;
				}
			}

			public override void onRegistrationCredentialsRequired(SinchClient client, ClientRegistration clientRegistration)
			{
			}
		}
	}

}